dll/win32/shell32/CDefView.cpp

```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
+* Column handling
+*/
+static HRESULT SHGetLVColumnSubItem(HWND List, UINT Col)
+{
+    LVCOLUMN lvc;
+    lvc.mask = LVCF_SUBITEM;
+    if (!ListView_GetColumn(List, Col, &lvc))
+        return E_FAIL;
+    else
+        return lvc.iSubItem;
+}
+
+HRESULT CDefView::MapFolderColumnToListColumn(UINT FoldCol)
+{
+    // This function is only called during column management, performance is not critical.
+    for (UINT i = 0;; ++i)
+    {
+        HRESULT r = SHGetLVColumnSubItem(m_ListView.m_hWnd, i);
+        if ((UINT) r == FoldCol)
+            return i;
+        else if (FAILED(r))
+            return r;
+    }
+}
+
+HRESULT CDefView::MapListColumnToFolderColumn(UINT ListCol)
+{
+    // This function is called every time a LVITEM::iSubItem is mapped to
+    // a folder column (calls to GetDetailsOf etc.) and should be fast.
+    if (m_ListToFolderColMap)
+    {
+        UINT count = DPA_GetPtrCount(m_ListToFolderColMap);
+        if (ListCol < count)
+        {
+            HRESULT col = (INT)(INT_PTR) DPA_FastGetPtr(m_ListToFolderColMap, ListCol);
+            assert(col >= 0 && col == SHGetLVColumnSubItem(m_ListView.m_hWnd, ListCol));
+            return col;
+        }
+        else if (count)
+        {
+            TRACE("m_ListToFolderColMap cache miss while mapping %d\n", ListCol);
+        }
+    }
+    return SHGetLVColumnSubItem(m_ListView.m_hWnd, ListCol);
+}
 
+HRESULT CDefView::GetDetailsByFolderColumn(PCUITEMID_CHILD pidl, UINT FoldCol, SHELLDETAILS &sd)
+{
+    // According to learn.microsoft.com/en-us/windows/win32/shell/sfvm-getdetailsof
+    // the query order is IShellFolder2, IShellDetails, SFVM_GETDETAILSOF.
+    HRESULT hr;
     if (m_pSF2Parent)
     {
-        for (int i = 0; 1; i++)
+        hr = m_pSF2Parent->GetDetailsOf(pidl, FoldCol, &sd);
+    }
+    if (FAILED(hr) && m_pSDParent)
+    {
+        hr = m_pSDParent->GetDetailsOf(pidl, FoldCol, &sd);
+    }
+#if 0 // TODO
+    if (FAILED(hr))
+    {
+        FIXME("Try SFVM_GETDETAILSOF\n");
+    }
+#endif
+    return hr;
+}
+
+HRESULT CDefView::GetDetailsByListColumn(PCUITEMID_CHILD pidl, UINT ListCol, SHELLDETAILS &sd)
+{
+    HRESULT hr = MapListColumnToFolderColumn(ListCol);
+    if (SUCCEEDED(hr))
+        return GetDetailsByFolderColumn(pidl, hr, sd);
+    ERR("Unable to determine folder column from list column %d\n", (int) ListCol);
+    return hr;
+}
+
+HRESULT CDefView::LoadColumn(UINT FoldCol, UINT ListCol, BOOL Insert)
+{
+    WCHAR buf[MAX_PATH];
+    SHELLDETAILS sd;
+    HRESULT hr;
+
+    sd.str.uType = !STRRET_WSTR; // Protect ourself from broken implementations
+    hr = GetDetailsByFolderColumn(NULL, FoldCol, sd);
+    if (FAILED(hr))
+        return hr;
+    hr = StrRetToStrNW(buf, _countof(buf), &sd.str, NULL);
+    if (FAILED(hr))
+        return hr;
+
+    LVCOLUMN lvc;
+    lvc.mask = LVCF_TEXT | LVCF_FMT | LVCF_WIDTH | LVCF_SUBITEM;
+    lvc.pszText = buf;
+    lvc.fmt = sd.fmt;
+    lvc.cx = sd.cxChar * 8; // FIXME: Font size? DPI?
```

Comment:
Might be outta scope but it would be nice to know what this is for sure?

Reply:
If I copy zipfldr.dll from XP to ROS, it reports 

```
Name:20
Type:15
Packed Size:10
Has a password:8
Size:8
Ratio:6
Date:25
```
for its `SHELLDETAILS::cxChar` values to our DefView. Mapping these to pixels with `TEXTMETRIC::tmAveCharWidth` does not match XP for whatever reason. `GetTextExtentPointW` gets very close however.

![XP and ROS](https://github.com/reactos/reactos/assets/138629473/3e3422f1-a23c-4b5b-93ea-2e826868efb3)